Using g++ and -Wmissing-declarations produces warning in generated C++ code

<!--  ❤️  -->
Steps tor reproduce: use g++ as compiler and take the tutorial and add
```
.flag("-Wmissing-declarations")
```
to `build.rs`. This produces the following warnings.
```
warning: /home/otto/cxx-demo/target/debug/build/cxx-demo-94d9b893a28bb988/out/cxxbridge/sources/cxx-demo/src/main.rs.cc:39:20: warning: no previous declaration for ‘BlobstoreClient* cxxbridge1$new_blobstore_client()’ [-Wmissing-declarations]
warning:    39 | ::BlobstoreClient *cxxbridge1$new_blobstore_client() noexcept {
warning:       |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /home/otto/cxx-demo/target/debug/build/cxx-demo-94d9b893a28bb988/out/cxxbridge/sources/cxx-demo/src/main.rs.cc:47:6: warning: no previous declaration for ‘void cxxbridge1$unique_ptr$BlobstoreClient$null(std::unique_ptr<BlobstoreClient>*)’ [-Wmissing-declarations]
warning:    47 | void cxxbridge1$unique_ptr$BlobstoreClient$null(::std::unique_ptr<::BlobstoreClient> *ptr) noexcept {
warning:       |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /home/otto/cxx-demo/target/debug/build/cxx-demo-94d9b893a28bb988/out/cxxbridge/sources/cxx-demo/src/main.rs.cc:50:6: warning: no previous declaration for ‘void cxxbridge1$unique_ptr$BlobstoreClient$raw(std::unique_ptr<BlobstoreClient>*, BlobstoreClient*)’ [-Wmissing-declarations]
warning:    50 | void cxxbridge1$unique_ptr$BlobstoreClient$raw(::std::unique_ptr<::BlobstoreClient> *ptr, ::BlobstoreClient *raw) noexcept {
warning:       |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /home/otto/cxx-demo/target/debug/build/cxx-demo-94d9b893a28bb988/out/cxxbridge/sources/cxx-demo/src/main.rs.cc:53:26: warning: no previous declaration for ‘const BlobstoreClient* cxxbridge1$unique_ptr$BlobstoreClient$get(const std::unique_ptr<BlobstoreClient>&)’ [-Wmissing-declarations]
warning:    53 | ::BlobstoreClient const *cxxbridge1$unique_ptr$BlobstoreClient$get(::std::unique_ptr<::BlobstoreClient> const &ptr) noexcept {
warning:       |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /home/otto/cxx-demo/target/debug/build/cxx-demo-94d9b893a28bb988/out/cxxbridge/sources/cxx-demo/src/main.rs.cc:56:20: warning: no previous declaration for ‘BlobstoreClient* cxxbridge1$unique_ptr$BlobstoreClient$release(std::unique_ptr<BlobstoreClient>&)’ [-Wmissing-declarations]
warning:    56 | ::BlobstoreClient *cxxbridge1$unique_ptr$BlobstoreClient$release(::std::unique_ptr<::BlobstoreClient> &ptr) noexcept {
warning:       |                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: /home/otto/cxx-demo/target/debug/build/cxx-demo-94d9b893a28bb988/out/cxxbridge/sources/cxx-demo/src/main.rs.cc:59:6: warning: no previous declaration for ‘void cxxbridge1$unique_ptr$BlobstoreClient$drop(std::unique_ptr<BlobstoreClient>*)’ [-Wmissing-declarations]
warning:    59 | void cxxbridge1$unique_ptr$BlobstoreClient$drop(::std::unique_ptr<::BlobstoreClient> *ptr) noexcept {
warning:       |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/gen/src/pragma.rs
+++ b/gen/src/pragma.rs
@@ -6,6 +6,7 @@ pub(crate) struct Pragma<'a> {
     pub gnu_diagnostic_ignore: BTreeSet<&'a str>,
     pub clang_diagnostic_ignore: BTreeSet<&'a str>,
     pub dollar_in_identifier: bool,
+    pub missing_declarations: bool,
     pub return_type_c_linkage: bool,
     pub begin: Content<'a>,
     pub end: Content<'a>,
@@ -19,9 +20,10 @@ impl<'a> Pragma<'a> {
 
 pub(super) fn write(out: &mut OutFile) {
     let Pragma {
-        ref gnu_diagnostic_ignore,
+        ref mut gnu_diagnostic_ignore,
         ref mut clang_diagnostic_ignore,
         dollar_in_identifier,
+        missing_declarations,
         return_type_c_linkage,
         ref mut begin,
         ref mut end,
@@ -30,9 +32,13 @@ pub(super) fn write(out: &mut OutFile) {
     if dollar_in_identifier {
         clang_diagnostic_ignore.insert("-Wdollar-in-identifier-extension");
     }
+    if missing_declarations {
+        gnu_diagnostic_ignore.insert("-Wmissing-declarations");
+    }
     if return_type_c_linkage {
         clang_diagnostic_ignore.insert("-Wreturn-type-c-linkage");
     }
+    let gnu_diagnostic_ignore = &*gnu_diagnostic_ignore;
     let clang_diagnostic_ignore = &*clang_diagnostic_ignore;
 
     if !gnu_diagnostic_ignore.is_empty() {

--- a/gen/src/write.rs
+++ b/gen/src/write.rs
@@ -576,6 +576,7 @@ fn write_struct_operator_decls<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
 
     if derive::contains(&strct.derives, Trait::PartialEq) {
         out.pragma.dollar_in_identifier = true;
+        out.pragma.missing_declarations = true;
         let link_name = mangle::operator(&strct.name, "eq");
         writeln!(
             out,
@@ -595,6 +596,7 @@ fn write_struct_operator_decls<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
 
     if derive::contains(&strct.derives, Trait::PartialOrd) {
         out.pragma.dollar_in_identifier = true;
+        out.pragma.missing_declarations = true;
         let link_name = mangle::operator(&strct.name, "lt");
         writeln!(
             out,
@@ -629,6 +631,7 @@ fn write_struct_operator_decls<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
     if derive::contains(&strct.derives, Trait::Hash) {
         out.include.cstddef = true;
         out.pragma.dollar_in_identifier = true;
+        out.pragma.missing_declarations = true;
         let link_name = mangle::operator(&strct.name, "hash");
         writeln!(
             out,
@@ -732,6 +735,7 @@ fn write_opaque_type_layout_decls<'a>(out: &mut OutFile<'a>, ety: &'a ExternType
     out.set_namespace(&ety.name.namespace);
     out.begin_block(Block::ExternC);
     out.pragma.dollar_in_identifier = true;
+    out.pragma.missing_declarations = true;
 
     let link_name = mangle::operator(&ety.name, "sizeof");
     writeln!(out, "::std::size_t {}() noexcept;", link_name);
@@ -779,6 +783,7 @@ fn begin_function_definition(out: &mut OutFile) {
 
 fn write_cxx_function_shim<'a>(out: &mut OutFile<'a>, efn: &'a ExternFn) {
     out.pragma.dollar_in_identifier = true;
+    out.pragma.missing_declarations = true;
     out.next_section();
     out.set_namespace(&efn.name.namespace);
     out.begin_block(Block::ExternC);
@@ -1761,6 +1766,7 @@ fn write_unique_ptr_common(out: &mut OutFile, ty: UniquePtr) {
     out.include.new = true;
     out.include.utility = true;
     out.pragma.dollar_in_identifier = true;
+    out.pragma.missing_declarations = true;
 
     let inner = ty.to_typename(out.types);
     let instance = ty.to_mangled(out.types);
@@ -1869,6 +1875,7 @@ fn write_shared_ptr(out: &mut OutFile, key: &NamedImplKey) {
     out.include.new = true;
     out.include.utility = true;
     out.pragma.dollar_in_identifier = true;
+    out.pragma.missing_declarations = true;
 
     // Some aliases are to opaque types; some are to trivial types. We can't
     // know at code generation time, so we generate both C++ and Rust side
@@ -1965,6 +1972,7 @@ fn write_weak_ptr(out: &mut OutFile, key: &NamedImplKey) {
     out.include.new = true;
     out.include.utility = true;
     out.pragma.dollar_in_identifier = true;
+    out.pragma.missing_declarations = true;
 
     writeln!(
         out,
@@ -2036,6 +2044,7 @@ fn write_cxx_vector(out: &mut OutFile, key: &NamedImplKey) {
     out.include.utility = true;
     out.builtin.destroy = true;
     out.pragma.dollar_in_identifier = true;
+    out.pragma.missing_declarations = true;
 
     begin_function_definition(out);
     writeln!(

--- a/src/cxx.cc
+++ b/src/cxx.cc
@@ -23,6 +23,7 @@
 #endif
 
 #ifdef __GNUC__
+#pragma GCC diagnostic ignored "-Wmissing-declarations"
 #pragma GCC diagnostic ignored "-Wshadow"
 #endif
 #ifdef __clang__

